packages/nimble-components/src/wafer-map/modules/rendering.ts

```diff
@@ -49,61 +64,76 @@ export class RenderingModule {
+        const dieWidth = this.wafermap.dataManager.dieDimensions.width;
+        const dieHeight = this.wafermap.dataManager.dieDimensions.height;
+        const transformedCanvasMinPoint = this.wafermap.transform.invert([
+            0, 0
+        ]);
+        const transformedCanvasMaxPoint = this.wafermap.transform.invert([
+            this.wafermap.canvas.width,
+            this.wafermap.canvas.height
+        ]);
+        transformedCanvasMinPoint[0] -= dieWidth;
+        transformedCanvasMinPoint[1] -= dieHeight;
 
         for (const die of this.dies) {
-            if (!prev) {
-                this.wafermap.canvasContext.fillStyle = die.fillStyle;
-            }
-            if (prev && die.fillStyle !== prev.fillStyle && die.fillStyle) {
-                this.wafermap.canvasContext.fillStyle = die.fillStyle;
+            if (
+                this.isDieVisible(
+                    die,
+                    transformedCanvasMinPoint,
+                    transformedCanvasMaxPoint
+                )
+            ) {
+                if (fillStyle !== die.fillStyle) {
+                    context.fillStyle = die.fillStyle;
+                    fillStyle = die.fillStyle;
+                }
+                context.fillRect(die.x, die.y, dieWidth, dieHeight);
             }
-            this.wafermap.canvasContext.fillRect(
-                die.x,
-                die.y,
-                this.wafermap.dataManager.dieDimensions.width,
-                this.wafermap.dataManager.dieDimensions.height
-            );
-            prev = die;
         }
     }
 
     private renderText(): void {
-        const dieSize = this.wafermap.dataManager.dieDimensions.width
-            * this.wafermap.dataManager.dieDimensions.height
-            * (this.wafermap.transform.k || 1);
-        const fontsize = this.wafermap.dataManager.labelsFontSize;
-        this.wafermap.canvasContext.font = `${fontsize.toString()}px sans-serif`;
-        this.wafermap.canvasContext.fillStyle = '#ffffff';
-        this.wafermap.canvasContext.textAlign = 'center';
-        this.wafermap.canvasContext.lineCap = 'butt';
-        const approxTextHeight = this.wafermap.canvasContext.measureText('M');
-
-        const dieDimensions = this.wafermap.dataManager.dieDimensions;
+        const dieWidth = this.wafermap.dataManager.dieDimensions.width;
+        const dieHeight = this.wafermap.dataManager.dieDimensions.height;
+        const dieSize = dieWidth * dieHeight * (this.wafermap.transform.k || 1);
         if (dieSize >= this.minDieDim) {
+            const fontsize = this.wafermap.dataManager.labelsFontSize;
+            const context = this.wafermap.canvasContext;
+            context.font = `${fontsize.toString()}px sans-serif`;
+            context.fillStyle = '#ffffff';
+            context.textAlign = 'center';
+            context.lineCap = 'butt';
+            const approxTextHeight = context.measureText('M');

```

Comment:
nit: Just go ahead and spell out `approximate`. The extra 5 characters is fine. :)